### PR TITLE
PackageQuery: Add `filter_extras`

### PR DESCRIPTION
**Pull request by m-blaha**
Filter installed "extra" packages - packages which are installed, but not available in any repository.

### CODE REVIEW

**Comment by j-mracek** [comment]
LGTM, but there is one request for a different implementation of extras filter. [Bug 1800910](https://bugzilla.redhat.com/show_bug.cgi?id=1800910) - dnf list extras stopped to shows some packages that should.
I think we should also provide that functionality, but I have no idea, how to delivery it for commandline.

**Comment by m-blaha** [comment]
Hm, yeah. The implementation in the patch uses the same approach as current dnf4 - extras are detected using `name.arch`.
Maybe adding new switch (and then also a filter), e.g. `--extras-versions` which can use algorithm that takes also package versions into account? What do you think?

**Comment by m-blaha** [comment]
I've added a commit which extends `filter_extras()` with optional `with_versions` parameter. If this is set to `true`, the whole NEVRA is used to detect extra packages.

File: include/libdnf/rpm/package_query.hpp [hunk 1]
```diff
@@ -587,8 +587,9 @@ class PackageQuery : public PackageSet {
     // TODO(dmach): consider removing the installed packages during the filtering
     void filter_priority();
 
-    /// Filter packages, which are installed but not available in any enabled repository
-    void filter_extras();
+    /// Filter packages, which are installed but not available in any enabled repository.
+    /// @param with_versions If false (default) extras calculation is based only on `name.arch`. That means package is not in extras if any version of the package exists in the repository. If true, filter_extras is more strict and returns each package which exact NEVRA is not present in any repository.
+    void filter_extras(const bool with_versions=false);
```
Comment (j-mracek) [inline include/libdnf/rpm/package_query.hpp:592]: Nice, technically it is not a version but EVR.

Just a suggestion. What about to name the variable as `exact_evr` or `exact_EVR`?

File: include/libdnf/rpm/package_query.hpp [hunk 2]
```diff
@@ -587,6 +587,10 @@ class PackageQuery : public PackageSet {
     // TODO(dmach): consider removing the installed packages during the filtering
     void filter_priority();
 
+    /// Filter packages, which are installed but not available in any enabled repository.
+    /// @param with_versions If false (default) extras calculation is based only on `name.arch`. That means package is not in extras if any version of the package exists in any of the enabled repositories. If true, filter_extras is more strict and returns each package which exact NEVRA is not present in any enabled repository.
```
Comment (j-mracek) [inline include/libdnf/rpm/package_query.hpp:591]: What about to replace `and returns each package which exact NEVRA is not present in any repository.` by `and keeps in query installed packages which exact NEVRA is not present in any repository.`

File: include/libdnf/rpm/package_query.hpp [hunk 3]
```diff
@@ -587,6 +587,10 @@ class PackageQuery : public PackageSet {
     // TODO(dmach): consider removing the installed packages during the filtering
     void filter_priority();
 
+    /// Filter packages, which are installed but not available in any enabled repository.
```
Comment (j-mracek) [inline include/libdnf/rpm/package_query.hpp:590]: What about to replace `Filter packages` by `Keep packages`?
May I also ask you to describe the behavior with regular excluded packages?

File: include/libdnf/rpm/package_query.hpp (hunk 1 above)
Comment (m-blaha) [reply include/libdnf/rpm/package_query.hpp]: Naming, the hardest part of the coding...
I agree, your suggested name for parameter is clearer.

File: include/libdnf/rpm/package_query.hpp (hunk 3 above)
Comment (m-blaha) [reply include/libdnf/rpm/package_query.hpp]: Sure. Done.

File: include/libdnf/rpm/package_query.hpp [hunk 4]
```diff
@@ -587,6 +587,16 @@ class PackageQuery : public PackageSet {
     // TODO(dmach): consider removing the installed packages during the filtering
     void filter_priority();
 
+    /// Keep in query only packages that are installed but not available in any enabled
+    /// repository. Also installed packages that are only part of non-active modules
+    /// are considered extras.
```
Comment (j-mracek) [inline include/libdnf/rpm/package_query.hpp:592]: Nice improvement, What about to `considered extras` change  to `considered as extras`?

File: libdnf/rpm/package_query.cpp [hunk 5]
```diff
@@ -2395,4 +2395,18 @@ void PackageQuery::filter_duplicates() {
...
+    filter_installed();
+    // Create query with available packages without non-modular excludes.
+    // As extras should be considered also packages in non-active modules.
+    PackageQuery available(p_impl->base, ExcludeFlags::IGNORE_REGULAR_EXCLUDES);
+    available.filter_available();
+    if (exact_evr) {
+        filter_nevra(available, sack::QueryCmp::NEQ);
+    } else {
+        filter_name_arch(available, sack::QueryCmp::NEQ);
+    }
```
Comment (j-mracek) [inline libdnf/rpm/package_query.cpp:2408]: I am really exited how easily you were able to implemented the filter.

File: include/libdnf/rpm/package_query.hpp (hunk 4 above)
Comment (m-blaha) [reply include/libdnf/rpm/package_query.hpp]: Done. Sorry for all those mistakes in my English.

File: include/libdnf/rpm/package_query.hpp [hunk 6]
```diff
@@ -587,6 +587,16 @@ class PackageQuery : public PackageSet {
     // TODO(dmach): consider removing the installed packages during the filtering
     void filter_priority();
 
+    /// Keep in query only packages that are installed but not available in any enabled
```
Comment (j-mracek) [inline include/libdnf/rpm/package_query.hpp:590]: May I ask you to add a note that filter ignores regular excludes?

**Comment by j-mracek** [comment]
LGTM, feel free to merge it when tests pass.

**Review by j-mracek — APPROVED** [review]

File: include/libdnf/rpm/package_query.hpp (hunk 6 above)
Comment (m-blaha) [reply include/libdnf/rpm/package_query.hpp]: oops. Sorry, I missed this comment. Should be fixed now.

**Pull request merged**